parametric/test_128_bit_traceids.py

```diff
@@ -0,0 +1,329 @@
+    "library_env", [{"DD_TRACE_PROPAGATION_STYLE": "Datadog", "DD_TRACE_128_BIT_TRACEID_GENERATION_ENABLED": "true",}],
+)
+def test_datadog_128_bit_generation_enabled(test_agent, test_library):
+    """Ensure that 128-bit TraceIds are properly generated, propagated in
+    datadog headers, and populated in trace data.
+    """
+    with test_library:
+        headers = make_single_request_and_get_inject_headers(test_library, [])
+    header_trace_id = headers["x-datadog-trace-id"]
+    span = get_span(test_agent)
+    span_trace_id = span.get("trace_id")
+    dd_p_tid = span["meta"].get("_dd.p.tid")
+
+    assert len(header_trace_id) < 21
+    assert int(header_trace_id, 10) == span_trace_id
+    validate_dd_p_tid(dd_p_tid)
+
+
+@pytest.mark.skip_library("dotnet", "not implemented")
+@pytest.mark.skip_library("golang", "not implemented")
+@pytest.mark.skip_library("java", "not implemented")
+@pytest.mark.skip_library("nodejs", "not implemented")
+@pytest.mark.skip_library("php", "Issue: traces not available from test agent")
+@pytest.mark.skip_library("python", "not implemented")
+@pytest.mark.skip_library("python_http", "not implemented")
+@pytest.mark.skip_library("ruby", "not implemented")
+@pytest.mark.parametrize(
+    "library_env",
+    [{"DD_TRACE_PROPAGATION_STYLE": "B3 single header", "DD_TRACE_128_BIT_TRACEID_GENERATION_ENABLED": "false",}],
+)
+def test_b3single_128_bit_propagation(test_agent, test_library):
+    """Ensure that external 128-bit TraceIds are properly propagated in B3
+    single-header.
+    """
+    with test_library:
+        headers = make_single_request_and_get_inject_headers(
+            test_library, [["b3", "640cfd8d00000000abcdefab12345678-000000003ade68b1-1"],],
+        )
+    span = get_span(test_agent)
+    trace_id = span.get("trace_id")
+    dd_p_tid = span["meta"].get("_dd.p.tid")
+    b3 = headers["b3"]
+
+    assert trace_id == int("abcdefab12345678", 16)
+    assert dd_p_tid == "640cfd8d00000000"
+    assert len(b3) > 32 + 16
+    check_128_bit_trace_id(b3[0:32], trace_id, dd_p_tid)
+
+
+@pytest.mark.skip_library("python", "Issue: Python doesn't pad the trace-id to length of 16 or 32 lower-hex characters")
+@pytest.mark.skip_library("java", "Issue: Java doesn't pad the trace-id to length of 16 or 32 lower-hex characters")
+@pytest.mark.skip_library("ruby", "Issue: Ruby doesn't support case-insensitive distributed headers")
+@pytest.mark.parametrize(
+    "library_env",
+    [{"DD_TRACE_PROPAGATION_STYLE": "B3 single header", "DD_TRACE_128_BIT_TRACEID_GENERATION_ENABLED": "false",}],
+)
+def test_b3single_128_bit_generation_disabled(test_agent, test_library):
+    """Ensure that 64-bit TraceIds are properly generated, propagated in B3
+    single-header, and populated in trace data.
+    """
+    with test_library:
+        headers = make_single_request_and_get_inject_headers(test_library, [])
+    span = get_span(test_agent)
+    b3 = headers["b3"]
+
+    assert len(b3) > 16 + 16
```

Comment:
This could be stricter and check for length == 16+16+3 (or 32+16+3 if we allow 64-bit trace ids to be zero-padded to 32 characters).

Reply:
I added some of the length tests for basic sanity checking but they are redundant wrt the existing propagation tests. I've updated these tests to verify the fields more precisely now rather than just checking length.